genesis/engine/couplers/sap_coupler.py

```diff
@@ -1079,119 +1623,298 @@ def compute_contact_regularization(self, sap_info, i_p, w_rms, time_step):
+        self.contact_candidate_type = ti.types.struct(
+            batch_idx=gs.ti_int,  # batch index
+            geom_idx=gs.ti_int,  # index of the FEM element
+            intersection_code=gs.ti_int,  # intersection code for the element
+            distance=gs.ti_vec4,  # distance vector for the element
+        )
+        self.n_contact_candidates = ti.field(gs.ti_int, shape=())
+        self.max_contact_candidates = self.fem_solver.n_surface_elements * self.fem_solver._B
+        self.contact_candidates = self.contact_candidate_type.field(shape=(self.max_contact_candidates,))
+
+        self.contact_pair_type = ti.types.struct(
+            batch_idx=gs.ti_int,  # batch index
+            geom_idx=gs.ti_int,  # index of the FEM element
+            barycentric=gs.ti_vec4,  # barycentric coordinates of the contact point
+            contact_pos=gs.ti_vec3,  # contact position
+            sap_info=self.sap_contact_info_type,  # contact info
+        )
+        self.max_contact_pairs = self.fem_solver.n_surface_elements * self.fem_solver._B
+        self.contact_pairs = self.contact_pair_type.field(shape=(self.max_contact_pairs,))
+
+    @ti.func
+    def detection(self, f: ti.i32):
+        overflow = False
+        candidates = ti.static(self.contact_candidates)
+        # Compute contact pairs
+        self.n_contact_candidates[None] = 0
+        # TODO Check surface element only instead of all elements
+        for i_b, i_e in ti.ndrange(self.fem_solver._B, self.fem_solver.n_elements):
+            intersection_code = ti.int32(0)
+            distance = ti.Vector.zero(gs.ti_float, 4)
+            for i in ti.static(range(4)):
+                i_v = self.fem_solver.elements_i[i_e].el2v[i]
+                pos_v = self.fem_solver.elements_v[f, i_v, i_b].pos
+                distance[i] = pos_v.z - self.fem_solver.floor_height
+                if distance[i] > 0.0:
+                    intersection_code |= 1 << i
+
+            # check if the element intersect with the floor
+            if intersection_code != 0 and intersection_code != 15:
+                i_c = ti.atomic_add(self.n_contact_candidates[None], 1)
+                if i_c < self.max_contact_candidates:
+                    candidates[i_c].batch_idx = i_b
+                    candidates[i_c].geom_idx = i_e
+                    candidates[i_c].intersection_code = intersection_code
+                    candidates[i_c].distance = distance
+                else:
+                    overflow = True
+
+        pairs = ti.static(self.contact_pairs)
+        sap_info = ti.static(pairs.sap_info)
+        self.n_contact_pairs[None] = 0
+        # Compute pair from candidates
+        for i_c in range(self.n_contact_candidates[None]):
+            candidate = candidates[i_c]
+            i_b = candidate.batch_idx
+            i_e = candidate.geom_idx
+            intersection_code = candidate.intersection_code
+            distance = candidate.distance
```

Comment:
Could you avoid this temporary and use `candidate.distance` directly? I think here you are creating a copy nor a reference